src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -245,25 +261,96 @@ private static String extractQualifiedName(DetailAST ast) {
     }
 
     /**
-     * Register to outer super classes of given classAst that
+     * Register to outer super class of given classAst that
      * given classAst is extending them.
      *
-     * @param classAst class which outer super classes will be
+     * @param classAst class which outer super class will be
      *                 informed about nesting subclass
+     * @param qualifiedClassName qualifies class name(with package) of the current class
      */
-    private void registerNestedSubclassToOuterSuperClasses(DetailAST classAst) {
+    private void registerNestedSubclassToOuterSuperClasses(DetailAST classAst,
+                                                           String qualifiedClassName) {
         final String currentAstSuperClassName = getSuperClassName(classAst);
         if (currentAstSuperClassName != null) {
-            for (ClassDesc classDesc : classes) {
-                final String classDescQualifiedName = classDesc.getQualifiedName();
-                if (doesNameInExtendMatchSuperClassName(classDescQualifiedName,
-                        currentAstSuperClassName)) {
+            final List<ClassDesc> sameNamedClassList =
+                    classesWithSameName(currentAstSuperClassName);
+            if (sameNamedClassList.isEmpty()) {
+                final ClassDesc classDesc = innerClasses.get(currentAstSuperClassName);
+                if (classDesc != null) {
                     classDesc.registerNestedSubclass();
                 }
             }
+            else {
+                registerNearestAsSuperClass(qualifiedClassName, sameNamedClassList);
+            }
         }
     }
 
+    /**
+     * Checks if there is a class with same name.
+     *
+     * @param className name of the class
+     * @return true if there is another class with same name.
+     * @noinspection CallToStringConcatCanBeReplacedByOperator
+     */
+    private List<ClassDesc> classesWithSameName(String className) {
+        return innerClasses.entrySet().stream()
+                .filter(entry -> {
+                    return entry.getKey().endsWith(
+                            PACKAGE_SEPARATOR.concat(className));
+                })
+                .map(Map.Entry::getValue)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * For all classes with the same name as the superclass, finds the nearest one
+     * and registers a subclass for it.
+     *
+     * @param superClassName current class name
+     * @param classesWithSameName classes which same name as super class
+     */
+    private static void registerNearestAsSuperClass(String superClassName,
+                                                    List<ClassDesc> classesWithSameName) {
+        final ClassDesc nearest = Collections.min(classesWithSameName, (first, second) -> {
```

Comment:
Just asking, but all this looks like a `compareTo` or `compare` method. We shouldn't be implementing that instead into `ClassDesc`?

Reply:
I'll give it a try. But I doubt that it is possible to define a strict order here.

Reply:
My fault I thought we were creating a simple comparator, but we are actually comparing an external object, `superClassName`.